test/rateLimiters/slidingWindowCounter.test.ts

```diff
@@ -0,0 +1,474 @@
+            afterEach(() => {
+                client.flushall();
+            });
+            test('fixed window is initially empty', async () => {
+                // window is intially empty
+                const withdraw5 = 5;
+                expect((await limiter.processRequest(user1, timestamp, withdraw5)).tokens).toBe(
+                    CAPACITY - withdraw5
+                );
+                const tokenCountFull = await getWindowFromClient(client, user1);
+                expect(tokenCountFull.currentTokens).toBe(CAPACITY - withdraw5);
+            });
+
+            test('fixed window is partially full and request has leftover tokens', async () => {
+                // Window is partially full but still has space for another small request
+                const initial = 6;
+                const partialWithdraw = 3;
+                expect((await limiter.processRequest(user2, timestamp, initial)).tokens).toBe(
+                    CAPACITY - initial
+                );
+                expect(
+                    (await limiter.processRequest(user2, timestamp, partialWithdraw)).tokens
+                ).toBe(CAPACITY - (initial + partialWithdraw));
+
+                const tokenCountPartial = await getWindowFromClient(client, user2);
+                expect(tokenCountPartial.currentTokens).toBe(
+                    CAPACITY - (initial + partialWithdraw)
+                );
+            });
+
+            // window partially full and no leftover tokens after request
+            test('fixed window is partially full and request has no leftover tokens', async () => {
+                const initial = 6;
+                await setTokenCountInClient(client, user2, initial, null, timestamp);
+                expect((await limiter.processRequest(user2, timestamp, initial)).tokens).toBe(0);
+                const tokenCountPartialToEmpty = await getWindowFromClient(client, user2);
+                expect(tokenCountPartialToEmpty.currentTokens).toBe(0);
+            });
+
+            // Bucket initially empty but enough time elapsed to paritally fill bucket since last request
+            test('fixed window is initially full but after new fixed window is initialized request is allowed', async () => {
+                await setTokenCountInClient(client, user4, 10, null, timestamp);
+                // tokens returned in processRequest is equal to the capacity
+                // still available in the fixed window
+                expect(
+                    (await limiter.processRequest(user4, timestamp + WINDOW_SIZE + 1, 10)).tokens
+                ).toBe(0);
+                // `currentTokens` cached is the amount of tokens
+                // currently in the fixed window.
+                // this differs from token bucket, which caches the amount
+                // of tokens still available for use
+                const count = await getWindowFromClient(client, user4);
+                expect(count.currentTokens).toBe(10);
```

Comment:
I think this would be zero because you just processed a request on line 95

Reply:
current tokens in the cache represents the amount of tokens in the current fixed window, unlike the cache of the token bucket, which represents the amount of tokens available to be used

Reply:
haha ya I need to change the way I'm thinking about it